### PR TITLE
dataset: pin validation shard 

### PR DESCRIPTION
**Pull request by KartikVashishta**
Addresses #541 Validation shard is fixed so val loss/bpb is stable regardless of how many shards were downloaded.

### CODE REVIEW

File: nanochat/dataset.py [hunk 1]
```diff
@@ -23,6 +23,12 @@
 BASE_URL = "https://huggingface.co/datasets/karpathy/fineweb-edu-100b-shuffle/resolve/main"
 MAX_SHARD = 1822 # the last datashard is shard_01822.parquet
 index_to_filename = lambda index: f"shard_{index:05d}.parquet" # format of the filenames
+
+# Always use a fixed shard for val so that the metrics don't depend
+# on how many shards are downloaded
+VAL_SHARD_INDEX = MAX_SHARD
```
Comment (svlandeg) [inline nanochat/dataset.py]: I personally wouldn't link these two variables to eachother, because the risk would be that in the future for some reason `MAX_SHARD` is changed into a different value without realising the consequences for the validation step. You may as well could do `VAL_SHARD_INDEX=1822`.

I think I'd personally just pick shard 0, which will always be downloaded anyway. This might reduce some if-else logic. The downside being that we won't be able to compare future evaluations with previous ones as soon as we'd merge this.

File: nanochat/dataloader.py [hunk 2]
```diff
@@ -32,10 +32,9 @@ def _document_batches(split, resume_state_dict, tokenizer_batch_size):
     """
     ddp, ddp_rank, ddp_local_rank, ddp_world_size = get_dist_info()
 
-    parquet_paths = list_parquet_files()
-    assert len(parquet_paths) != 0, "No dataset parquet files found, did you run dataset.py?"
-    parquet_paths = parquet_paths[:-1] if split == "train" else parquet_paths[-1:]
 
+    parquet_paths = get_parquet_paths(split) 
+    assert len(parquet_paths) != 0, "No dataset parquet files found, did tiy run dataset.py?"
```
Comment (svlandeg) [inline nanochat/dataloader.py]:
```suggestion
    assert len(parquet_paths) != 0, "No dataset parquet files found, did you run dataset.py?"
```

File: nanochat/dataset.py [hunk 3]
```diff
@@ -40,15 +46,32 @@ def list_parquet_files(data_dir=None):
...
+    data_dir = DATA_DIR if data_dir is None else data_dir
+
+    val_path = os.path.join(data_dir, VAL_SHARD_FILENAME)
+    if split == "val":
+        if not os.path.exists(val_path):
+            raise FileNotFoundError(f"Validation shard {VAL_SHARD_FILENAME} not found in {data_dir}")
+        return [val_path]
+
+    # train everything except the val shard (if it exists, exclude it if not, nothing to exclude)
+    all_paths = list_parquet_files(data_dir)
```
Comment (svlandeg) [inline nanochat/dataset.py]: I think we can just merge these two functions and c/p the functionality of `list_parquet_files` here inline. That also allows you to prevent adding the val shard to the set instead of having to remove it again later on.

File: nanochat/dataset.py (hunk 1 above)
Comment (KartikVashishta) [reply nanochat/dataset.py]: good point - i've decoupled these. `VAL_SHARD_INDEX` is now pinned to 1822. I’m keeping it at shard_01822 specifically for backwards comparability.

File: nanochat/dataset.py (hunk 3 above)
Comment (KartikVashishta) [reply nanochat/dataset.py]: Agree with the intent. I kept list_parquet_files() as a small utility, but added exclude_filenames so get_parquet_paths("train") never includes the val shard in the first place

File: nanochat/dataset.py [hunk 4]
```diff
@@ -23,32 +23,58 @@
...
+    val_path = os.path.join(data_dir, VAL_SHARD_FILENAME)
+    if split == "val":
+        if not os.path.exists(val_path):
+            raise FileNotFoundError(
+                f"Validation shard {VAL_SHARD_FILENAME} not found in {data_dir}. "
+                f"Run: python -m nanochat.dataset -n <N> (downloads the val shard too)."
+            )
+        return [val_path]
+    else:
+        #train split: list files while excluding val (don't add then remove).
```
Comment (svlandeg) [inline nanochat/dataset.py]:
```suggestion
        # train split: list files while excluding val
```

File: nanochat/dataset.py [hunk 5]
```diff
@@ -111,13 +137,21 @@ def download_single_file(index):
 
 if __name__ == "__main__":
     parser = argparse.ArgumentParser(description="Download FineWeb-Edu 100BT dataset shards")
-    parser.add_argument("-n", "--num-files", type=int, default=-1, help="Number of shards to download (default: -1), -1 = disable")
+    parser.add_argument("-n", "--num-files", type=int, default=-1, help="Number of training shards to download (default: -1 = all). Validation shard is always included")
```
Comment (svlandeg) [inline nanochat/dataset.py]: I would remove this comment because it may actually be confusing to users
```suggestion
    parser.add_argument("-n", "--num-files", type=int, default=-1, help="Number of training shards to download (default: -1 = all).")
```

File: nanochat/dataset.py [hunk 6]
```diff
@@ -111,13 +137,21 @@ def download_single_file(index):
...
     parser.add_argument("-w", "--num-workers", type=int, default=4, help="Number of parallel download workers (default: 4)")
     args = parser.parse_args()
 
     num = MAX_SHARD + 1 if args.num_files == -1 else min(args.num_files, MAX_SHARD + 1)
     ids_to_download = list(range(num))
-    print(f"Downloading {len(ids_to_download)} shards using {args.num_workers} workers...")
+    if VAL_SHARD_INDEX not in ids_to_download:
+        ids_to_download.append(VAL_SHARD_INDEX)
+    ids_to_download = sorted(set(ids_to_download))
```
Comment (svlandeg) [inline nanochat/dataset.py]: `set` or `sorted` shouldn't be necessary if we're not in the `if` condition, because `ids_to_download = list(range(num))` is just fine. Within this `if`, we also don't need to do `set` because we already checked `not in`.
```suggestion
        ids_to_download.append(VAL_SHARD_INDEX)
        ids_to_download = sorted(ids_to_download)
```

**Comment by KartikVashishta** [comment]
Thanks for the review @svlandeg !

File: nanochat/dataset.py (hunk 6 above)
Comment (svlandeg) [reply nanochat/dataset.py]: @KartikVashishta : can you address this comment?

Comment (KartikVashishta) [reply nanochat/dataset.py]: Yep, agreed - removed the set/sorted. range(num) is already unique and ordered, and we only append val when it’s not present

**Review by svlandeg — COMMENTED** [review]
This has now been fixed by a recent [edit](https://github.com/karpathy/nanochat/commit/324e69c45d3606095adb6b409078647145165454) on `master`, so I'll go ahead and close this one 🙏
